File: frontend/src/app/services/special-offer.service.ts
import {Injectable} from '@angular/core';
import {Globals} from "../global/globals";
import {HttpClient, HttpParams, HttpResponse} from "@angular/common/http";
import {Observable} from "rxjs";
import {SpecialOfferCreateDto, SpecialOfferDetailDto, SpecialOfferListDto} from "../dtos/special-offer";
import {formatIsoDate} from "../util/date-helper";

@Injectable({
  providedIn: 'root'
})
export class SpecialOfferService {
  private placeBaseUri: string = this.globals.backendUri + "/special-offers";


  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Get all special offers without image
   *
   * @return all special offers as a ListDto
   */
  getSpecialOffers(): Observable<SpecialOfferListDto[]> {
    return this.httpClient.get<SpecialOfferListDto[]>(this.placeBaseUri);
  }

  /**
   * Create a new special offer
   *
   * @param specialOfferCreateDto the special offer to create
   * @return an Observable for the HttpResponse
   */
  createSpecialOffer(specialOfferCreateDto: SpecialOfferCreateDto): Observable<SpecialOfferCreateDto> {
    const formData: FormData = new FormData();
    formData.append('name', specialOfferCreateDto.name);
    formData.append('pricePerPax', specialOfferCreateDto.pricePerPax.toString());
    formData.append('image', specialOfferCreateDto.image);
    return this.httpClient.post<SpecialOfferCreateDto>(this.placeBaseUri, formData);
  }

  /**
   * Delete a special offer by its ID
   *
   * @param id
   * @return an Observable for the HttpResponse
   */
  delete(id: number): Observable<HttpResponse<void>> {
    return this.httpClient.delete<void>(this.placeBaseUri, {observe: 'response', body: id});
  }

  /**
   * Get a special offer by its ID
   *
   * @param id
   * @return the special offer as a DetailDto
   */
  getSpecialOffer(id: number): Observable<SpecialOfferDetailDto> {
    return this.httpClient.get<SpecialOfferDetailDto>(this.placeBaseUri + "/detail", {params: new HttpParams().set('id', id)});
  }


  /**
   * Get all special offers with image
   *
   * @return all special offers as a DetailDto
   */
  getAllSpecialOffersWithDetail(): Observable<SpecialOfferDetailDto[]> {
    return this.httpClient.get<SpecialOfferDetailDto[]>(this.placeBaseUri + "/details");
  }
}

File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/SpecialOfferService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.SpecialOfferCreateDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.SpecialOfferDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.SpecialOfferListDto;

import java.util.List;

/**
 * Service for operations on special offers (e.g. creating a new special offer).
 */
public interface SpecialOfferService {

    /**
     * Creates a new special offer.
     *
     * @param specialOfferCreateDto the special offer to create
     * @return the created special offer
     */
    SpecialOfferCreateDto createSpecialOffer(SpecialOfferCreateDto specialOfferCreateDto);

    /**
     * Gets all special offers.
     *
     * @return all special offers
     */
    List<SpecialOfferListDto> getAllSpecialOffers();

    /**
     * Deletes a special offer.
     *
     * @param id the id of the special offer to delete
     */
    void deleteSpecialOffer(Long id);

    /**
     * Gets a special offer.
     *
     * @param id the id of the special offer to get
     * @return the special offer
     */
    SpecialOfferDetailDto getSpecialOffer(Long id);

    /**
     * Gets all special offers with image.
     *
     * @return all special offers as detailDTOs
     */
    List<SpecialOfferDetailDto> getAllSpecialOfferDetails();
}
